Add base-filesystem defines for arm64/ppc64el/riscv64 and make build quiet again

keszybz opened the pull request.

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,47 @@ static const BaseFilesystem table[] = {
 #else
-        /* gcc doesn't allow pragma to be used within constructs, hence log about this separately below */
-#  define WARN_LIB64 1
+#  warning "Uknown architecture"
```
typo

keszybz replied:
C'mon, it's the English language, the home of every possible spelling that doesn't make sense ;)

keszybz replied:
I dropped this line now. We'll still get the warning from the `#pragma message` below.

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,47 @@ static const BaseFilesystem table[] = {
                          "usr/lib64\0",                "ld-linux-x86-64.so.2" },
+#  define HAVE_LIB64_DIRS 1
+#elif defined(__ia64__)
+#elif defined(__m68k__)
```
so we know for some archs that they never had lib64/lib32 because there never was any other flavour of the CPU for a different register width. m68k is one of them, hence we can probably already skip that arch here.

This is interesting btw:

https://sourceware.org/glibc/wiki/ABIList
https://wiki.debian.org/Multiarch/LibraryPathOverview

According to those docs, it's pretty much these archs that need the symlink in some form for either lib32 or lib64: mips, x86/x86-64, ppc/ppc64, sparc.

poettering replied:
Those docs aren't complete apparently though, as #14311 suggests, which asks for this on aarch64 and s390x apparently. Maybe fedora actually insists on lib64 and thus /lib must be a symlink to it? yikes...

poettering replied:
So the ELF ABI specs dictate /lib64/ld-lsb-s390x.so.3 is the right name for ld.so on s390x:

https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwi07YbSxaH0AhVdSPEDHfo4AZ4QFnoECB0QAQ&url=https%3A%2F%2Frefspecs.linuxfoundation.org%2FLSB_3.2.0%2FLSB-Core-S390X%2FLSB-Core-S390X.pdf&usg=AOvVaw046SYJRRVtNCLApXjobCPX

keszybz replied:
I added an exclude fro m68k now. I'm not confident enough to add any other excludes.

keszybz replied:
Yeah, but that document is from 2007. After looking at the debian docs, which get the defines wrong (or at least different than what I see on an actual Fedora system), I'd be wary to add defines without a system to verify them against. People have access to s390x, so hopefully they'll fill this in. With the scaffolding in place, it should be fairly easy.

keszybz replied:
OK, with the information in https://github.com/systemd/systemd/issues/14311, we can be pretty certain about the paths. I added a commit for this now.

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -46,6 +46,9 @@ static const BaseFilesystem table[] = {
          * multiarch logic, but support Fedora-style multilib too.*/
 #if defined(__aarch64__)
+        { "lib64",    0, "usr/lib/aarch64-linux-gnu\0"
+                         "usr/lib64\0",                "ld-linux-aarch64.so.1" },
```
Hmm, now I am puzzled. The Debian doc you linked suggests lib64 is not used there? but on fedora it is?

But either way ld-linux-aarch64.so.1 is not in lib64 but lin lib hence not the right thing to look for here. i.e. it appears there should not be any conditionalization on ld.so existance here.

keszybz replied:
> but on fedora it is?

The directory is there. I included the listing in the commit message so if somebody wants to figure out why we added that, they'll know we didn't make it up ;)

The symlink is also there: `/lib64/ld-linux-aarch64.so.1 -> ld-2.32.so`.

poettering replied:
the point I was making is that I am not sure we need the lib64 symlink at all on this arch. I mean, in an ideal world the root fs would have no such symlinks at all. But unfortunately the ELF ABI of some archs declares that the ld.so dynamic loader is in /lib64/ hence if you want to be able to run any standards conforming dynamic ELF binary you have to have the lib64 symlink in place.

But, aarch64 is not one like that afaics, the official ld.so path for it is actually /lib/ld-linux-aarch64.so.1, not /lib64/ld-linux-aarch64.so.1, thus the necessity to have the /lib64 symlink in place is just not there.

For any other .so but the dynamic loader .so itself it's entirely sufficient that it appears in the default .so search path. The dynamic linker one is different however, it's path is hardcoded in the ELF header and part of the architecture ABI. Thus, for regular .so files we can just rely that whatever the shared libs are located below /lib the search logic will find it, only for the dynamic loader .so itself we need to make the symlink. Hence, I am questioning if we need the above at all...

poettering replied:
(but maybe I missing something because #14311 actually talks about /lib64/ specifically, hence I am a bit confused)

keszybz replied:
I think it's better to create the symlink, if it is already present in "normal" installations. It might not matter for the linker, but people occasionally use paths, and a path with `/lib64/` would normally work.

poettering replied:
but that means we'd never have "legacy-free" systems that don't bother with /lib64/ at all, we'd always carry that cruft around...

I mean, we could certainly do that, but if that's the idea then looking for the right ld.so file in that dir only makes vague sense, because the 4th field in the struct is about that: find the right dir below /usr/ that has the ld.so in it, because we need to hook it up. But if we don't actually need to hook it up, then we don't have to insist in it (and most likely should link /lib64 to /lib probably)

keszybz replied:
Oh, I checked the symlinks more, and it's actually the same situation as described in #14311. ` /lib/ld-linux-aarch64.so.1 -> ../lib64/ld-2.32.so`, so we **must** link `/lib64 → usr/lib64`. The patch was correct, but I'll extend the description a bit.

poettering replied:
hmm, ok, if fedora put their dynamic loader into lib64 even though the ABI defined lib for it, and then they fixed it up with a symlink, then I guess it indeed is the rght approach to symlink /lib64 to /usr/lib64 too, and also the check for the dynamic loader does indeed make sense.

Could you add a quick comment to the sources about this? Something like this:

```c
/* aarch64 ELF ABI actually says dynamic loader is in /lib/, but Fedora puts it in /lib64/ anyway and just symlinks /lib/ld-linux-aarch64.so.1 to ../lib64/ld-linux-aarch64.so.1. For this to work correctly /lib64/ must be symlinked to /usr/lib64/, hence do this in any case. */
```

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -78,6 +78,9 @@ static const BaseFilesystem table[] = {
 #  if __riscv_xlen == 32
 #  elif __riscv_xlen == 64
+        { "lib64",    0, "usr/lib/riscv64-linux-gnu\0"
+                         "usr/lib64\0",                "ld-linux-riscv64-lp64d.so.1" },
```
here too, the conditionalization on the existence of the ld.so doesn't look right given that it's in lib, not in lib64

keszybz replied:
`/lib64/ld-linux-riscv64-lp64d.so.1 -> ld-2.32.so`

poettering replied:
but /lib64/ld-linux-riscv64-lp64d.so doesn't really matter, does it? i.e. ELF binaries will list /lib/ld-linux-riscv64-lp64d.so in their ELF headers as dynamic loader, notthing else, hence /lib64 doesn't really have to be there as a symlink

poettering replied:
mayba then add a comment here too to the sources: `/* same situation as for aarch64 */`

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -66,6 +66,15 @@ static const BaseFilesystem table[] = {
 #elif defined(__powerpc__)
+#  if defined(__PPC64__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        { "lib64",    0, "usr/lib/powerpc64le-linux-gnu\0"
+                         "usr/lib64\0",                "ld64.so.2" },
```
this one does look superficially correct to me

poettering commented on src/shared/gpt.c:
```diff
@@ -4,16 +4,6 @@
-        !defined(GPT_USR_NATIVE_VERITY) ||                              \
-        !defined(GPT_USR_NATIVE_VERITY_SIG)
-#pragma message "Please define GPT partition types for your architecture."
-#endif
```
hmm? please see:

 https://github.com/systemd/systemd/pull/21290/commits/1fb2d8fcb69bcdbab0a5dd23bbf02f729e47e656

for the reasons i added this.

poettering replied:
i.e. the idea is that RootImage=, portable services, systemd-dissect, systemd-nspawn should should just work on every arch, and not just the ones that also do UEFI. the concept of GPT, portable services, containers are all arch independent, hence we should support them on all archs.

Yes, systed-gpt-auto-generator doesn't make too much sense on archs where the rootfs is not on GPT, but all other tools that parse GPT can certainly tkae benefit of it.

keszybz commented:
@xen0n please check commit "docs: fix bit-flip in LoongArch root partition UUID".

poettering commented on src/shared/gpt.h:
```diff
@@ -7,22 +7,38 @@
 #include "id128-util.h"
 
+#define GPT_ROOT_ALPHA       SD_ID128_MAKE(65,23,f8,ae,3e,b1,4e,2a,a0,5a,18,b6,95,ae,65,6f)
+#define GPT_ROOT_ARC         SD_ID128_MAKE(d2,7f,46,ed,29,19,4c,b8,bd,25,95,31,f3,c1,65,34)
```
hmm, our ARCHITECTURE_xyz enum knows ARC_BE as well? And PPC_LE? and a couple of others? we might want to keep that list in sync

keszybz replied:
I *can* add them, but I think I think they are not really used? If Debian doesn't have a triplet for them, then probably the hardware isn't available.

poettering commented on src/shared/gpt.h:
```diff
@@ -7,22 +7,38 @@
 #define GPT_ROOT_LOONGARCH64 SD_ID128_MAKE(77,05,58,00,79,2c,4f,94,b3,9a,99,c9,1b,76,2b,b6)
+#define GPT_ROOT_POWERPC     SD_ID128_MAKE(1d,e3,f1,ef,fa,98,47,b5,8d,cd,4a,86,0a,65,4d,78)
+#define GPT_ROOT_POWERPC64   SD_ID128_MAKE(91,2a,de,1d,a8,39,49,13,89,64,a1,0e,ee,08,fb,d2)
+#define GPT_ROOT_POWERPC64LE SD_ID128_MAKE(c3,1c,45,e6,3f,39,41,2e,80,fb,48,09,c4,98,05,99)
```
the ARCHITECTURE_XYZ enum calls these PPC btw, maybe stick to the same naming?

keszybz replied:
Done.

poettering commented:
(maybe even have a test that ensures that for each ARCHITECTURE_xyz enum we have a root-<suffix>, usr-<suffix> entry and so on, to ensure that whenever someone adds a new arch they immediately add the uuid, too)

xen0n commented:
> @xen0n please check commit "docs: fix bit-flip in LoongArch root partition UUID".

Hi, sorry for the inconvenience caused; this might be caused by my cat playing with the keyboard, with Vim running in the foreground and me not noticing :facepalm:

As some downstream projects are already using the UUID as in the docs, most notably [util-linux](https://github.com/util-linux/util-linux/commit/2d29fccaad267d1b003dc0ed2bb9634ff76f3e49#diff-44e8c98872fad5f41b4f9f2733a11180b6dd73f084503be37ef8f6ab95683e9c), we should adjust code to match doc; no shipping LoongArch systems are using upstream systemd so the change should break no one. This also matches the work being done by @yetist at loongarch64/systemd#7.

keszybz commented:
Rebased. Thank you for all the comments, I think it's looking pretty good now.

>  i think for riscv64 and aarch64 we can get away to not create the /lib64/ symlink.

As mentioned, we actually need the symlink in all cases. I added more commentary to the patches.

> Also, I'd really like if the arch names in architecture.h would match the GPT names

I renamed POWERPC → PPC.

> and that we have one GPT type for each arch we know there and vice versa.

I didn't do this part: I think some of those architectures are completely dead. I don't think we should ever *remove* defines, but I think it's unlikely that people will be ever creating those file systems. The table is already quite long…

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -60,6 +60,8 @@ static const BaseFilesystem table[] = {
 #elif defined(__ia64__)
 #elif defined(__m68k__)
+        /* No link needed. */
+#  define HAVE_LIB64_DIRS 1
```
maybe call the macro `KNOW_LIB64_DIRS`?

poettering commented:
looks all great, just would like to see a brief version of the commit msgs being added as comments to the sources too

keszybz commented:
Updated with the comments and macro rename. No functional change.

poettering commented:
lgtm

poettering commented:
Posted a follow-up PR in #21444, ptal.

poettering commented:
when using #21444 on top of #21425 one can see that this PR is missing a bunch of entries in gpt_partition_type_table[] for the newly defined type uuids.

(Feel free to just pull the commits of my PR into yours)

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,73 @@ static const BaseFilesystem table[] = {
+        /* aarch64 ELF ABI actually says dynamic loader is in /lib/, but Fedora puts it in /lib64/ anyway and
+         * just symlinks /lib/ld-linux-aarch64.so.1 to ../lib64/ld-linux-aarch64.so.1. For this to work
+         * correctly, /lib64/ must be symlinked to /usr/lib64/. */
+        { "lib64",    0, "usr/lib/aarch64-linux-gnu\0"
```
one more idea: replace the string at the right with: `"usr/lib/" LIB_ARCH_TUPLE "\0"`

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,73 @@ static const BaseFilesystem table[] = {
+        /* No /lib64 on arm. The linker is /lib/ld-linux-armhf.so.3. */
+#  define KNOW_LIB64_DIRS 1
+#elif defined(__i386__) || defined(__x86_64__)
         { "lib64",    0, "usr/lib/x86_64-linux-gnu\0"
```
same here

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,73 @@ static const BaseFilesystem table[] = {
+#  endif
+#elif defined(__powerpc__)
+#  if defined(__PPC64__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        { "lib64",    0, "usr/lib/powerpc64le-linux-gnu\0"
```
ditto

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,73 @@ static const BaseFilesystem table[] = {
+#  if __riscv_xlen == 32
+#  elif __riscv_xlen == 64
+        /* Same situation as for aarch64 */
+        { "lib64",    0, "usr/lib/riscv64-linux-gnu\0"
```
ditto

poettering commented on src/shared/base-filesystem.c:
```diff
@@ -38,25 +38,73 @@ static const BaseFilesystem table[] = {
+        /* s390-linux-gnu */
+#elif defined(__s390x__)
+        { "lib64",    0, "usr/lib/s390x-linux-gnu\0"
+                         "usr/lib64",                  "ld-lsb-s390x.so.3" },
```
ditto

keszybz commented:
Your test was very useful. I didn't know that there are so many different places to adjust ;(
I change the code to use more macros to reduce the amount of manual work requires when adding new arches. I think it's easier to keep in sync now.

I added the GPT defines also for mipsel and mips64el. Those architectures are listed as active ports by Debian, so I guess it makes sense to have them here too. I didn't add anything else.

poettering commented on src/shared/gpt.c:
```diff
@@ -14,87 +14,55 @@
-        { GPT_TMP,                         "tmp"                         },
-        { GPT_USER_HOME,                   "user-home"                   },
-        { GPT_LINUX_GENERIC,               "linux-generic"               },
+        _GPT_ARCH_SEXTET(SECONDARY,   "secondary"),
```
hmm this should require #ifdeffery, no? i mean only some archs have a secondary arch?

keszybz replied:
Oh, indeed. I don't know why I dropped the the `#ifdef`.

poettering commented on src/basic/architecture.c:
```diff
@@ -118,20 +118,19 @@ int uname_architecture(void) {
         assert_se(uname(&u) >= 0);
 
-        for (i = 0; i < ELEMENTSOF(arch_map); i++)
+        for (unsigned i = 0; i < ELEMENTSOF(arch_map); i++)
```
size_t?

keszybz replied:
Considering that this is a hand-written map, I think it's unlikely we'll go above 4e9 entries ;) But I'll change it.
